vulkan_generator/codegen_utils/codegen.py

```diff
@@ -0,0 +1,180 @@
+
+    ret = return_type + " " + name + "("
+
+    for arg_name, arg_type in arguments.items():
+        ret = ret + arg_type + " " + arg_name + ", "
+
+    # Strip the trailing comma from the last argument
+    if arguments:
+        ret = ret[0:-2]
+
+    ret = ret + ")"
+
+    return ret
+
+def create_function_declaration(name: str,
+                                return_type: str = "void",
+                                arguments: Optional[Dict[str, str]] = None) -> str:
+    return create_function_signature(name, return_type, arguments,) + ";"
+
+def create_function_definition(name: str,
+                               return_type: str = "void",
+                               arguments: Optional[Dict[str, str]] = None,
+                               code: str = "") -> str:
+    return create_function_signature(name, return_type, arguments) + " {\n" + indent_code(code) +"\n}"
+
+def create_exception_declaration(name: str, base_class: str = "std::exception") -> str:
+    return "class " + name + " : public " + base_class +" {};"
+
+def create_class_definition(name: str,
+                            public_inheritance: Optional[List[str]] = None,
+                            protected_inheritance: Optional[List[str]] = None,
+                            private_inheritance: Optional[List[str]] = None,
+                            public_members: Optional[List[str]] = None,
+                            protected_members: Optional[List[str]] = None,
+                            private_members: Optional[List[str]] = None,
+                            public_functions: Optional[List[str]] = None,
+                            protected_functions: Optional[List[str]] = None,
+                            private_functions: Optional[List[str]] = None
+                            ) -> str:
+
+    def inheritance_helper(inheritance_type : str, inheritance : List[str]) -> str:
+        if not inheritance:
+            return ""
+        inheritance_string = inheritance_type + " "
+        for base in inheritance:
+            inheritance_string = inheritance_string + base + ", "
+        return inheritance_string
+
+    def member_helper(member_type : str, members : List[str]) -> str:
+        if not members:
+            return ""
+        members_string = indent() + member_type +":\n"
+        for member in members:
+            members_string = members_string + indent(2) + member + "\n"
+        return members_string + "\n"
+
+    ret = "class " + name
+
+    inheritance = public_inheritance or protected_inheritance or private_inheritance
+    if inheritance:
+        ret = ret +" : "
+
+    ret = ret +inheritance_helper("public", public_inheritance)
+    ret = ret +inheritance_helper("protected", protected_inheritance)
+    ret = ret +inheritance_helper("private", private_inheritance)
+
+    # Strip the trailing comma from the last inherited base class
```

Comment:
The code actually becomes more complicated if we don't do this here. Otherwise for example the public inheritance needs to worry about whether there will be any private or protected inheritance following it when deciding if it should add or not add a comma at the end.